Initial blog post (#1)
ambaxter/bbolt-rs: blog.md

```
 % ./target/release/bench -w seq -r seq
# Write 64.945474ms     (64.945µs/op)   (15397 op/sec)
# Read  1.000046609s    (24ns/op)       (41666666 op/sec)
```

> With those two lines the last few months of labor bear fruit. 

Reviewer: ```suggestion
With those results, the last few months of labor bear fruit. 
```

> I reimplemented the BBolt database in Rust. 

Reviewer: ```suggestion
Over those months, I have reimplemented the Go-based [BBolt key-value database](https://github.com/etcd-io/bbolt) in Rust. 
```

> It is the culmination of my journey where I learned how to build a database, how to design database architecture in Rust, and how to use API design to remove footguns.

Reviewer: ```suggestion
This marks the culmination of my journey where I learned how to:
* Build a database
* Design a Rust-centric database architecturet
* How to use API design to remove ["footguns"](https://en.wiktionary.org/wiki/footgun)
```

# Impetus

> The project started last year when I wanted to learn how ETCD and the Raft algorithm worked which then lead me to learning how the BBolt database worked. 

Reviewer: ```suggestion
The project started in late 2023 when I wanted to learn how ETCD and the Raft algorithm work.
This also led me to learn how the BBolt database works. 
```

> Following in the software engineering tradition what started as curiosity turned into an all-consuming case of yak shaving. 

Reviewer: ```suggestion
I proceeded to follow in the software-engineering tradition of allowing curiosity to turn into an all-consuming case of [yak shaving](https://en.wiktionary.org/wiki/yak_shaving). 
```

> 

Reviewer: I would expand a little more on the impetus here beyond just "idle curiosity"/"I was bored" lol. Specifically, which were the technical challenges that piqued your curiosity that produced the yak shaving?

> # Goals and Standards

Reviewer: ```suggestion
# Project Standards
```

* Built in stable Rust
* Library API as close to the original as possible
* On-file representation exactly matches original code 
* Manage transaction memory without Rc

> * Unsafe behavior validated by Miri

Reviewer: ```suggestion
* Use only stable Rust
* Keep the library API as close to the original as possible
* The on-file representation _must_ match original code exactly
* Manage transaction memory without Rc
* Unsafe behavior validated by Miri
```
Links to Rc and Miri?

> # BBolt Under a Microscope

Reviewer: I really like this heading.

> Under the covers BBolt is a minimalistic B+ tree database designed to store Buckets. 

Reviewer: ```suggestion
Under the covers, BBolt is a minimalistic B+ tree database designed to store buckets. 
```
If BBolt has a named object called "Bucket", I would instead maybe stylize it as `Bucket` (surround with backticks).

> Buckets may store key/value pairs in bytes and sub-Buckets. 

Reviewer: ```suggestion
_Buckets_ store key-value pairs in bytes and sub-Buckets. 
```
Again, consider backticks.

> Sub-buckets may be stored inline inside a parent bucket or on its own page.  

Reviewer: ```suggestion
Sub-buckets live inline within a parent bucket or on their own page.
```
It might be better to flip this around and make "parent bucket" the subject in order to more-easily avoid passive tense.

It's unclear (to me) what a page means here.

> The data types used to represent the BBolt database are relatively straightforward. 

Reviewer: ```suggestion
The data types that represent the BBolt database are relatively straightforward.
For example, the following <thing> <verbs> [a <another thing>].
```

I'm not sure saying it is straightforward tells the reader anything. It's either a self-evident fact or it runs the risk that, if they _don't_ find it straightforward, that the rest of the article is going to be over their head and they should quit reading.

> ```rust

Reviewer: I would break this up into multiple code snippets with fewer inline comments and more plain-English explanations of what's going on. Don't feel like you need to explain all of Rust, but don't assume that your audience are all Rust experts.

Also, if there are key parts of this code (perhaps the bigger structs?), then maybe only include and talk about those. You can handwave away a few things here if it improves overall clarity.

/// The Page ID. Page address = `PgId` * meta.page_size
#[repr(C)]
pub struct PgId(u64);

/// The Transaction ID. Monotonic and incremented every commit
#[repr(C)]
pub struct TxId(u64);

> /// `PageHeader` represents the on-file layout of a page header. 

Reviewer: This is a prime example. I would _hope_ that something called a "page header" represents a "page header" lol. What is a page header? Why do I care about it?

/// 
/// `page` in Go BBolt
#[repr(C)]
pub struct PageHeader {
  /// This Page's ID
  id: PgId,
  /// Page's type. Branch(0x01), Leaf(0x02), Meta(0x04), or FreeList(0x10)
  flags: u16,
  /// Defines the number of items in the Branch, Leaf, and Freelist pages
  count: u16,
  /// How many additional meta.page_size pages are included in this page
  overflow: u32,
}

> /// `BucketHeader` represents the on-file layout of a bucket header.

Reviewer: This is a bit better. I would pull this out of a codeblock and do a little more explanation in plain English.

Author: These are direct copies of the struct definitions. Would it not be better to leave the descriptions as they are in the code?

  version: u32,
  /// Database page size where page address = [PgId] * meta.page_size
  page_size: u32,
  flags: u32,
  /// Root bucket header
  root: BucketHeader,
  /// FreeList page location
  free_list: PgId,
  /// The end of the database where EOF = meta.pgid * meta.page_size
  pgid: PgId,
  /// Current transaction ID
  txid: TxId,
  /// Checksum of the previous Meta fields using the 64-bit version of the Fowler-Noll-Vo hash function
  checksum: u64,
}

/// `LeafPageElement` represents the on-file layout of a leaf page's element
/// 
/// `leafPageElement` in Go BBolt 
#[repr(C)]
pub struct LeafPageElement {
  /// Additional flag for each element. If leaf is a Bucket then 0x01 set
  flags: u32,
  /// The distance from this element's pointer to its key/value location
  pos: u32,
  /// Key length
  key_size: u32,
  /// Value length
  value_size: u32,
}

///`BranchPageElement` represents the on-file layout of a branch page's element
/// 
/// `branchPageElement` in Go BBolt
#[repr(C)]
pub struct BranchPageElement {
  /// The distance from this element's pointer to its key location
  pos: u32,
  /// Key length
  key_size: u32,
  /// Page ID of this branch's
  pgid: PgId,
}

```

> The 4 database page types are also straightforward. 

Reviewer: Similar as before. Don't necessarily assume _all_ of your audience is intimately familiar with database architecture.

Here, look to provide explanations that would answer questions like: what is a page? Why are there different types? Why do I care about those types.

Page Layouts
```

>       Meta Page

Reviewer: Similar to the codeblock before, break each of these (really nice) diagrams into separate codeblocks and provide a blurb for each, with transitions/captions.

The goal is to be able to skim the article and get an idea of what's contained therein. Naturally, if I see largely two big codeblocks with very little English, I'm inclined to just assume it's a dump of some code and lose interest.

      ├──────┤
      │ Meta │
      └──────┘

   Freelist Page
      ┌──────┐
      │ Page │
      │Header│
      ├──────┤     First element will be the FreeList
      │PgId 0│◀────   count if PageHeader.count ==
      ├──────┤                  U16::MAX
      │PgId 1│
      ├──────┤
      │PgId 2│
      ├──────┤
      │ .... │
      └──────┘

 Leaf Element Page
      ┌──────┐
      │ Page │
      │Header│
      ├──────┤
      │Elem 0│
      ├──────┤
      │Elem 1│       key ptr =
      ├──────┤  elem ptr + elem.pos
      │Elem 2│   ┌─── ┌───┐
      ├──────┤   │    │ k │
      │ .... │   │    │ e │
      ├──────┤───┘    │ y │
      │K/V 0 │        │ v │
      ├──────┤───┐    │ a │
      │K/V 1 │   │    │ l │
      ├──────┤   │    │ u │
      │K/V 2 │   │    │ e │
      ├──────┤   └─── └───┘
      │ .... │
      └──────┘

Branch Element Page
      ┌──────┐
      │ Page │
      │Header│
      ├──────┤
      │Elem 0│
      ├──────┤
      │Elem 1│
      ├──────┤
      │Elem 2│
      ├──────┤
      │ .... │
      ├──────┤           key ptr =
      │Key 0 │◀──── elem ptr + elem.pos
      ├──────┤
      │Key 1 │
      ├──────┤
      │Key 2 │
      ├──────┤
      │ .... │
      └──────┘

> Created with Monodraw

Reviewer: Good attribution, but pull it out of the codeblocks and/or provide the info in footer. It's essentially tertiary information.

```

> # Foundations of Reading Mmapped Files

Reviewer: Good heading, though I do wonder if this section should be higher. I also immediately wonder "what is an mmapped file and why do I care?" You basically have to constantly provide reasons for your reader to keep reading.

Author: The mmapped file and why do I care is definitely coming. I got pulled into a meeting before I could get too much further here

[memmap2](https://crates.io/crates/memmap2)

> [Mmap](https://en.wikipedia.org/wiki/Mmap) [CreateFileMappingW](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-createfilemappingw)

Reviewer: Good attributions! I would have a `# References` section at the bottom with a lot of these types of links though (including Monodraw).

Author: I will be using these links in my description of mmap

> ## Reading a Page

Reviewer: Is "reading" the best verb here? Who is doing the reading?

## Reading a Meta Page

> ```go

Reviewer: You probably know what I'm going to say here :) English explanations!

// meta returns a pointer to the metadata section of the page.
func (p *page) meta() *meta {
	return (*meta)(unsafeAdd(unsafe.Pointer(p), unsafe.Sizeof(*p)))
}
```

## Reading INodes

```go
// leafPageElement retrieves the leaf node by index
func (p *page) leafPageElement(index uint16) *leafPageElement {
	return (*leafPageElement)(unsafeIndex(unsafe.Pointer(p), unsafe.Sizeof(*p),
		leafPageElementSize, int(index)))
}
```

```go
// key returns a byte slice of the node key.
func (n *leafPageElement) key() []byte {
	i := int(n.pos)
	j := i + int(n.ksize)
	return unsafeByteSlice(unsafe.Pointer(n), 0, i, j)
}
```

> # Removing Footguns With API Design

Reviewer: ```suggestion
# Hardening the API at the Design Level
```

I would reserve using cutesy jargon terms like "footgun" in the paragraphs themselves, not headers. Again, assume most readers are going to skim before reading.
